### PR TITLE
[NVPTX] Fix code generation for `trap-unreachable`.

### PR DESCRIPTION
**Pull request by chsigg**
https://reviews.llvm.org/D152789 added an `exit` op before each `unreachable`. This means we never get to the `trap` instruction.

This change limits the insertion of `exit` instructions to the cases where `unreachable` is not lowered to `trap`. Trap itself is changed to be emitted as `trap; exit;` to convey to `ptxas` that it exits the CFG.

### CODE REVIEW

**Comment by chsigg** [comment]
> I think we would then want skip lowering ISD::TRAP, so that we don't end up with trap; exit; trap; sequences.

Actually, no. Because `ISD::TRAP` can also be generated elsewhere, not just from `unreachable` (e.g. from `Intrinsic::trap`).

**Comment by chsigg** [comment]
I moved the implementation to `NVPTXLowerUnreachablePass`, which is better because it also fixes the `no-trap-after-noreturn` case.

I'm wondering though if we should convey to `ptxas` that `trap` exits the CFG by always emitting `ISD::TRAP` as `trap; exit;` (under the assumption that we don't support predicated trap). This would avoid the slightly weird `trap; exit; trap;` for `trap-unreachable` (i.e., `NVPTXLowerUnreachablePass` would only add `exit` iff `unsupported` is dropped later on).

**Comment by Artem-B** [comment]
> always emitting ISD::TRAP as `trap; exit;`

That would be fine.

**Comment by Artem-B** [comment]
@maleadt FYI. For some reason I can't add you as a reviewer on this PR.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 1]
```diff
@@ -99,6 +105,24 @@ char NVPTXLowerUnreachable::ID = 1;
...
+  return "add an exit instruction before every unreachable";
+}
+
+// =============================================================================
+// Returns whether a `trap` intrinsic should be emitted before I.
+//
+// This is a copy of the logic in SelectionDAGBuilder::visitUnreachable().
+// =============================================================================
+bool NVPTXLowerUnreachable::shouldEmitTrap(const UnreachableInst &I) const {
+  if (!TrapUnreachable)
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:118]: Nit: `if (!(TrapUnreachable || NoTrapAfterNoreturn))`

**Review by Artem-B — COMMENTED** [review]
LGTM in general.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 1 above)
Comment (chsigg) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:118]: I'm not sure I understand. `NoTrapAfterNoreturn` has no effect if `TrapUnreachable == false`.

**Comment by chsigg** [comment]
> > always emitting ISD::TRAP as `trap; exit;`
> 
> That would be fine.

Ok, I changed it to that. It seems a little cleaner because it a) avoids `trap; exit; trap;` and b) conveys to `ptxas` that `trap`s from intrinsics exit the CFG as well.

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp [hunk 2]
```diff
@@ -3226,14 +3226,9 @@ void SelectionDAGBuilder::visitUnreachable(const UnreachableInst &I) {
 
   // We may be able to ignore unreachable behind a noreturn call.
   if (DAG.getTarget().Options.NoTrapAfterNoreturn) {
-    const BasicBlock &BB = *I.getParent();
-    if (&I != &BB.front()) {
-      BasicBlock::const_iterator PredI =
-        std::prev(BasicBlock::const_iterator(&I));
```
Comment (Artem-B) [inline llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3232]: What was the purpose of the removed code?  Was it removed intentionally?

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 1 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:118]: I'm suggesting collapsing two early exits into one. Sorry, I should've marked that the comment applies to more than one line. I'm still figuring out pull request review mechanics.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp [hunk 3]
```diff
@@ -99,12 +105,33 @@ char NVPTXLowerUnreachable::ID = 1;
...
+}
+
 // =============================================================================
 // Main function for this pass.
 // =============================================================================
 bool NVPTXLowerUnreachable::runOnFunction(Function &F) {
   if (skipFunction(F))
     return false;
+  // Early out iff isLoweredToTrap() always returns true.
+  if (TrapUnreachable && !NoTrapAfterNoreturn)
```
Comment (Artem-B) [inline llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:133]: Nit: `!NoTrapAfterNoReturn` has disturbingly high concentration of negations. :-/
Can we rename it into `TrapAfterNoReturn` ?

File: llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp (hunk 2 above)
Comment (chsigg) [reply llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:3232]: The functionality of the new code is the same, it's just much easier to read.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 1 above)
Comment (chsigg) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:118]:
```
if (!(TrapUnreachable && NoTrapAfterNoreturn))
  return TrapUnreachable;
```
... seems unnecessarily complex to save two LOCs, no?

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 3 above)
Comment (chsigg) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:133]: Fully agree, I can do that as a follow-up.

**Review by maleadt — APPROVED** [review]
LGTM.

It didn't occur to me to preserve the `trap` as in Julia we're mostly avoiding that instruction since it 'breaks' the GPU for the current process, without a way to reset.

EDIT: just to be sure, I ran our GPU testsuite on an older device where https://github.com/JuliaGPU/CUDA.jl/issues/1746 is an issue, and using `trap; exit` as proposed here doesn't seem to cause any issues.

File: llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp (hunk 1 above)
Comment (Artem-B) [reply llvm/lib/Target/NVPTX/NVPTXLowerUnreachable.cpp:118]: LLVM style tends to favor brevity, so if multiple exits do not buy us anything in terms of functionality or readability, fewer lines wins. This way you can see more code without having to scroll up/down, which does help with larger functions.

In this case it's fine either way as the function is small.

**Review by Artem-B — APPROVED** [review]

**Comment by Artem-B** [comment]
We've removed `nvptx-exit-on-unreachable` knob which used to disable generation of `exit`.

https://github.com/llvm/llvm-project/pull/72641 tries to disable it for some unreachable branches, which I'm not convinced we should do, but we do not seem to have an escape hatch option any more, as an alternative.

**Pull request merged**